options/internal/generic/essential.cpp

```diff
@@ -199,9 +199,8 @@ void *memmove(void *dest, const void *src, size_t size) {
 	char *dest_bytes = (char *)dest;
 	char *src_bytes = (char *)src;
 	if(dest_bytes < src_bytes) {
-		for(size_t i = 0; i < size; i++)
-			dest_bytes[i] = src_bytes[i];
-	}else{
+		memcpy(dest, src, size);
```

Comment:
I don't think that this is correct; `memcpy` does not make any guarantees that the data will be copied forwards and not backwards. This branch however only ensures that a forwards copy is safe, which is why the for loop was in place previously.

Reply:
More specifically, memcopying overlapping regions is UB and is a classic bug that has lead to numerous vulnerabilities.

Reply:
but mlibc also provides `memcpy`, so would it not be fine to rely on it's inner workings due to the fact we provide it to begin with?

Reply:
Yes, this is a non-issue since we also provide memcpy() and we build with `-fno-builtin`

Reply:
Yes, we control memcopy, but again we do not guarantee how it operates. For example, if we were to provide Arch-specific optimized versions (e.g. using SIMD) this could reasonably be broken. As-is, this is just incorrect.

Reply:
> Yes, we control memcopy, but again we do not guarantee how it operates. For example, if we were to provide Arch-specific optimized versions (e.g. using SIMD) this could reasonably be broken. As-is, this is just incorrect.

"This is just incorrect" is an overly harsh statement. As is, it is correct in that it works given our only implementation of `memcpy()` that exists. It could *potentially* become incorrect in the future if our implementation of `memcpy()` changes such that the assumption of copy direction is broken.

Until then, the most that is needed is a comment stating how `memmove()` depends on the copy direction of `memcpy()` and to be careful in case the latter is changed.

Reply:
This is not a "technically correct" or "technically incorrect" a-ha gotcha discussion given *we* are the libc so even my usage of the word "assumption" is not really correct, since we *know* how `memcpy()` works by virtue of the fact that *we are* the ones implementing it.